lib/AST/USRGeneration.cpp

```diff
@@ -135,8 +135,10 @@ bool ide::printDeclUSR(const ValueDecl *D, raw_ostream &OS) {
   } else if (isa<TypeAliasDecl>(VD) || isa<AssociatedTypeDecl>(VD)) {
     Mangler.mangleContextOf(VD);
     Mangler.mangleDeclName(VD);
-  } else {
+  } else if (!VD->isInvalid()) {
     Mangler.mangleEntity(VD, /*uncurryingLevel=*/0);
```

Comment:
The problem is that "valid decl" is a loosely-defined concept. Ideally, validateDecl() would always produce a decl with a correct generic environment, a name that can be mangled, etc regardless of any syntax errors in the user's code.

The problem with checking isInvalid() is two-fold:

- If your checks are overly-conservative, you'll miss opportunities to provide more diagnostics to the user.

- If you forget a check somewhere, you might still crash.

I hope this makes sense -- even if the user's code is malformed, our AST should maintain as many invariants as possible to simplify downstream code, like IDE support.